services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DelegateDecoupledAuthenticationProvider.java

```diff
@@ -0,0 +1,281 @@
+        // which can bind authResultId with Decoupled Authentication Result of Authentication by AD.
+        // By including client_id, Decoupled Authentication Callback Endpoint can recognize the CD(Consumption Device) who sent Backchannel Authentication Request.
+
+        // The following scopes should be displayed on AD(Authentication Device):
+        // 1. scopes specified explicitly as query parameter in the authorization request
+        // 2. scopes specified implicitly as default client scope in keycloak
+        CIBALoginUserResolver resolver = session.getProvider(CIBALoginUserResolver.class);
+        if (resolver == null) {
+            throw new RuntimeException("CIBA Login User Resolver not setup properly.");
+        }
+        String authRequestedUserHint = realm.getCIBAPolicy().getAuthRequestedUserHint();
+        UserModel user = null;
+        if (authRequestedUserHint.equals(CIBAConstants.LOGIN_HINT)) {
+            user = resolver.getUserFromLoginHint(request.getLoginHint());
+        } else if (authRequestedUserHint.equals(CIBAConstants.ID_TOKEN_HINT)) {
+            user = resolver.getUserFromLoginHint(request.getIdTokenHint());
+        } else if (authRequestedUserHint.equals(CIBAConstants.LOGIN_HINT_TOKEN)) {
+            user = resolver.getUserFromLoginHint(request.getLoginHintToken());
+        } else {
+            throw new RuntimeException("CIBA invalid Authentication Requested User Hint.");
+        }
+        String infoUsedByAuthentication = resolver.getInfoUsedByAuthentication(user);
+
+        StringBuilder scopeBuilder = new StringBuilder();
+        Map<String, ClientScopeModel> defaultScopesMap = client.getClientScopes(true, true);
+        defaultScopesMap.forEach((key, value)->{if (value.isDisplayOnConsentScreen()) scopeBuilder.append(value.getName()).append(" ");});
+        String defaultClientScope = scopeBuilder.toString();
+
+        CIBAAuthReqId decoupledAuthIdJwt = new CIBAAuthReqId();
+        decoupledAuthIdJwt.id(KeycloakModelUtils.generateId());
+        decoupledAuthIdJwt.setScope(request.getScope());
+        decoupledAuthIdJwt.setSessionState(userSessionIdWillBeCreated);
+        decoupledAuthIdJwt.setAuthResultId(authResultId);
+        decoupledAuthIdJwt.issuedNow();
+        decoupledAuthIdJwt.issuer(Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+        decoupledAuthIdJwt.audience(decoupledAuthIdJwt.getIssuer());
+        decoupledAuthIdJwt.subject(user.getId());
+        decoupledAuthIdJwt.exp(Long.valueOf(Time.currentTime() + expiresIn));
+        //decoupledAuthIdJwt.issuedFor(Decoupled Auth Server's client_id); TODO
+        decoupledAuthIdJwt.issuedFor(client.getClientId()); // TODO : set CD's client_id intentionally, not Decoupled Auth Server. It is not good idea so that client_id field should be added.
```

Comment:
I agree it is not a good idea.

IMO, having the authentication channel as a client in Keycloak is probably the best way to go. We could potentially use different authentication methods when sending requests to the service too.

We could even send signed/encrypted requests.

The way it is right now is not much secure, from the service perspective.

One more reason, I think, to make have a `HttpAuthenticationChannel` so that we can use different forms of HTTP Authentication/Authorization, even if only BASIC.

Reply:
@pedroigor I see, my understanding is that we had better to introduce the concept "Authentication Channel" just above whose role is to communicate between keycloak and the decouple authentication server (but not limited to) and provide several its implementation like `HTTPAuthenticationChannel` that can provide several security level (e.g. transport layer : HTTP only,  HTTPS  / client authentication : BASIC, X.509). Is that right?

Reply:
Yeah. However, I won't push for introducing these authentication aspects right now if you think that it should be enough to document HTTPS between Keycloak and the authentication channel.

Let's see what @mposolda thinks.